### ISSUE TITLE
Index out of range when processing some JavaScript files

### ISSUE THREAD

**Issue by EdCharbeneau** [dotnet/templating#1170]
I'm creating a template based on the template from the JavaScript Services Angular template. This template has two files main-server.js and main-client.js. The presence of either of these .js files in the template package cause a failure when running the template.

Example: 
running `dotnet new kendo-ng` results in the following error.

> Template "MVC ASP.NET Core with Kendo UI for Angular" could not be created.
> Error while processing file /content/kendo-ng/ClientApp/dist/main-server.js
> Index was out of range. Must be non-negative and less than the size of the collection.
> Parameter name: index

I've included the .nupkg I'm using below to help recreate the issue.

https://1drv.ms/u/s!Ak8aQRebrTvWou4qMpO_Pcfo8-4HjA

**Comment by mlorbetske** [comment]
On lines 112149 and 112393 of main-server.js, there are `//#if` directives whose expression isn't parenthesized which will cause issues with the current C-style expression evaluator. If these aren't intended for processing, you can mark these files are `copyOnly` as @sayedihashimi has suggested or put `//-:cnd` at the top of the file to turn off processing of conditionals ( @seancpeters should be able to provide more info on how to do either of these if needed)

**Comment by EdCharbeneau** [comment]
Where is `copyOnly` set? Modifying the files isn't ideal since they are code generated files.

**Comment by sayedihashimi** [comment]
Hi @EdCharbeneau `copyOnly` is set in the `template.json` file. See the docs at:

 - https://github.com/dotnet/templating/wiki/Reference-for-template.json
 - https://github.com/dotnet/templating/wiki/%22Runnable-Project%22-Templates

**Comment by EdCharbeneau** [comment]
`copyOnly` isn't working. running `dotnet new kendo-ng` never completes.

Since these files are created from webpack, they are not 100% necessary to include in the template. They do however create a better out-of-the-box developer experience because it requires less knowledge by the end user to get started (ie: running webpack commands). I'm looking into excluding the files completely and using npm to resolve the DX. However, I wonder if others will run into this issue.

**Comment by seancpeters** [comment]
@EdCharbeneau can you copy paste the `copyOnly` configuration that you've set up? That functionality works in general, with the config I can help troubleshoot why its not working.

**Comment by EdCharbeneau** [comment]
@seancpeters 

```
      "copyOnly": [
        "ClientApp/dist/main-server.js",
        "wwwroot/dist/main-client.js"
      ]
```

**Comment by EdCharbeneau** [comment]
Revisiting my earlier comment about using npm: These files can be excluded from the template completely and instead I've added an `install` script to package.json that triggers the files to be recreated. 

```
  "scripts": {
    "test": "karma start ClientApp/test/karma.conf.js",
    "install": "npm run webpack:vendor && npm run webpack",
    "webpack:vendor": "webpack --config webpack.config.vendor.js",
    "webpack": "webpack"
  },
```
This reduces the .nupkg file size and bypasses the issue. 

My concern is now with helping correct any tooling issues that may exist.

**Comment by seancpeters** [comment]
I modified a local copy of your template, adding the copyOnly information, and it worked correctly. Two possibilities come to mind as to why it's not working for you:
1) There were some bugs with copyOnly in older versions of dotnet new. If you're running an old version that could be the problem. The version can be checked via `dotnet --info`. If you can copy-paste the version info, I can let you know if you're running a version prior to the copyOnly fix.
2) The copyOnly information is not in the right location in template.json. I couldn't tell from the pasted copyOnly info, but it should be part of "sources" - like this:
```
  "sources": [
    {
      "source": "./",
      "target": "./",
      "exclude": [
        ".template.config/**"
      ],
	  "copyOnly": [
        "ClientApp/dist/main-server.js",
        "wwwroot/dist/main-client.js"
      ],
    }
  ],
```

**Comment by EdCharbeneau** [comment]
1.) Verified, yes I have the correct `copyOnly` code in my template.

2.)

Running `dotnet --info` gives me `.NET Command Line Tools (2.0.0-preview2-006497)`

The template project in the nupkg linked above is still on `1.0.4`

I created a new version with everything updated to 2.0.0-preview2 and created a new nupkg. Now I receive an error: 
> Template "MVC ASP.NET Core with Kendo UI for Angular" could not be created.
Error while processing file /content/kendo-ng/WebApplicationBasic.csproj
Specified argument was out of the range of valid values.
Parameter name: offset

**Comment by seancpeters** [comment]
I'm having trouble reproducing the issue with the WebApplicationBasic.csproj file - the nupkg from the original link https://1drv.ms/u/s!Ak8aQRebrTvWou4qMpO_Pcfo8-4HjA doesn't contain a file with that name. The only .csproj file in the nupkg is KuiKitchenSink.csproj - have you made additional changes to your local copy that are not in the .nupkg in the link? 

Here's what I see in the content directory inside the nupkg:
![kendo-ng-content](https://user-images.githubusercontent.com/20912881/28844758-94ed59e6-76ba-11e7-89ec-460d550634d9.PNG)

**Comment by seancpeters** [comment]
With the updated package, I'm able to repro your issue. The bug causing the issue will be fixed in 2.1.0, but for now, to avoid the problem, we can turn off conditional processing in the .csproj file by adding this line at the top of the file:
`<!--/-:cnd -->`
This causes template processing to not process conditionals in that file - which is where the bug is.

To speed up the creation of the template, I'd recommend configuring the entire wwwroot directory to be copyonly, since those are large files and they don't have any template processing setup in them. This can be done with a glob in the copyOnly configuration, like this:
```
"copyOnly": [
         "ClientApp/dist/main-server.js",
         "wwwroot/**"
       ],
```
The template should still work without that change, but creation will be slower.

**Comment by EdCharbeneau** [comment]
> The bug causing the issue will be fixed in 2.1.0

That's great news. Thanks for your help on this.
